### PR TITLE
HPCC-15750 Implement getLikelihood for And, OR, NOT, optimize nested likely

### PR DESCRIPTION
**Pull request by shamser**

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday Please can you review

File: ecl/hql/hqlfold.cpp [hunk 1]
```diff
@@ -5183,8 +5183,12 @@ IHqlExpression * CExprFolderTransformer::doFoldTransformed(IHqlExpression * unfo
     case no_unlikely:
         {
             IHqlExpression * child = expr->queryChild(0);
-            if (child->queryValue())
-                return LINK(child);
+
+            if (child->getOperator()==no_likely || child->getOperator()==no_unlikely)
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: A few personal style comments.  Others may not agree
- if the true branch of an if finishes with a return, I don't generally include the else - I find it easier to read 
- an if inside an else clause, I would normally place it in curly brackets {}
- I would probably use a switch statement instead with this new case added.  (child->queryValue() test can be replaced with no_constant).
But none would stop me merging.

File: ecl/hql/hqlfold.cpp [hunk 2]
```diff
@@ -5183,8 +5183,12 @@ IHqlExpression * CExprFolderTransformer::doFoldTransformed(IHqlExpression * unfo
...
+                return createValue(op, expr->getType(), LINK(child->queryChild(0)));
```
Comment (ghalliday) [inline ecl/hql/hqlfold.cpp]: @shamser Hasn't this lost the likelihood associated with expr?  I would have expected

`replaceChild(expr, 0, child->queryChild(0));`

File: ecl/regress/likely2.ecl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+person := dataset('person', person_layout, thor);
+
+isLessXpos1000 := LIKELY(person.xpos < 1000, 0.1);
+isIdLess4 := LIKELY(person.person_id < 4,0.2);
+filtered0 := person( LIKELY(isLessXpos1000, 0.2) );
+filtered1 :=  person( isLessXpos1000 AND isIdLess4 );
+filtered2 :=  person( isLessXpos1000 OR isIdLess4 );
+filtered3 :=  person( NOT isLessXpos1000 );
+
+output(filtered1,,'tst_likely2.d01');
```
Comment (ghalliday) [inline ecl/regress/likely2.ecl]: filtered0 isn't included in the test, so it doesn't catch the issue above.

**Comment by ghalliday** [comment]
@shamser very close.

**Comment by shamser** [comment]
@ghalliday Please can you have another look?

**Pull request merged**